crates/red_knot_python_semantic/src/semantic_model.rs

```diff
@@ -0,0 +1,183 @@
+impl_expression_has_ty!(ast::ExprAttribute);
+impl_expression_has_ty!(ast::ExprSubscript);
+impl_expression_has_ty!(ast::ExprStarred);
+impl_expression_has_ty!(ast::ExprName);
+impl_expression_has_ty!(ast::ExprList);
+impl_expression_has_ty!(ast::ExprTuple);
+impl_expression_has_ty!(ast::ExprSlice);
+impl_expression_has_ty!(ast::ExprIpyEscapeCommand);
+
+impl HasTy for ast::Expr {
+    fn ty<'db>(&self, model: &SemanticModel<'db>) -> Type<'db> {
+        match self {
+            Expr::BoolOp(inner) => inner.ty(model),
+            Expr::Named(inner) => inner.ty(model),
+            Expr::BinOp(inner) => inner.ty(model),
+            Expr::UnaryOp(inner) => inner.ty(model),
+            Expr::Lambda(inner) => inner.ty(model),
+            Expr::If(inner) => inner.ty(model),
+            Expr::Dict(inner) => inner.ty(model),
+            Expr::Set(inner) => inner.ty(model),
+            Expr::ListComp(inner) => inner.ty(model),
+            Expr::SetComp(inner) => inner.ty(model),
+            Expr::DictComp(inner) => inner.ty(model),
+            Expr::Generator(inner) => inner.ty(model),
+            Expr::Await(inner) => inner.ty(model),
+            Expr::Yield(inner) => inner.ty(model),
+            Expr::YieldFrom(inner) => inner.ty(model),
+            Expr::Compare(inner) => inner.ty(model),
+            Expr::Call(inner) => inner.ty(model),
+            Expr::FString(inner) => inner.ty(model),
+            Expr::StringLiteral(inner) => inner.ty(model),
+            Expr::BytesLiteral(inner) => inner.ty(model),
+            Expr::NumberLiteral(inner) => inner.ty(model),
+            Expr::BooleanLiteral(inner) => inner.ty(model),
+            Expr::NoneLiteral(inner) => inner.ty(model),
+            Expr::EllipsisLiteral(inner) => inner.ty(model),
+            Expr::Attribute(inner) => inner.ty(model),
+            Expr::Subscript(inner) => inner.ty(model),
+            Expr::Starred(inner) => inner.ty(model),
+            Expr::Name(inner) => inner.ty(model),
+            Expr::List(inner) => inner.ty(model),
+            Expr::Tuple(inner) => inner.ty(model),
+            Expr::Slice(inner) => inner.ty(model),
+            Expr::IpyEscapeCommand(inner) => inner.ty(model),
+        }
+    }
+}
+
+impl HasTy for ast::StmtFunctionDef {
+    fn ty<'db>(&self, model: &SemanticModel<'db>) -> Type<'db> {
+        let index = semantic_index(model.db, model.file);
+        let definition_scope = index.definition_scope(NodeWithScopeKey::from(self));
+
+        // SAFETY: A function always has either an enclosing module, function or class scope.
+        let mut parent_scope_id = index.parent_scope_id(definition_scope).unwrap();
+        let parent_scope = index.scope(parent_scope_id);
+
+        if parent_scope.kind() == ScopeKind::Annotation {
+            parent_scope_id = index.parent_scope_id(parent_scope_id).unwrap();
+        }
+
+        let scope = parent_scope_id.to_scope_id(model.db, model.file);
```

Comment:
It feels weird that we have to go the inner scope of the function, then walk up one or two scopes, worrying about details like AnnotationScope, just to find the scope in which the FunctionDef node occurs. We already keep a map from node to scope it occurs in for expressions, and there are a lot more of those; maybe we should keep the same mapping for some statements, too?

Reply:
I agree that the code here is painful but I'm worried about tracking too much data over just recomputing when necessary. 

An alternative I considered that could also make things more consistent is that the `Definition -> Scope` map doesn't track the scope the node introduces, but the scope in which they are defined. This way, the scope returned is the same for all nodes, which isn't the case today where for functions, it's the function scope and for imports, it's the enclosing scope.

Reply:
I don't think the enclosing scope is enough information to find the correct inner scope if there are eg multiple classes with the same name defined in a scope. I think in the end we may want to track both separately for the cases where each can be relevant.

Reply:
I agree that it wouldn't add much. But performance dies by 100 cuts. I would like to avoid tracking redundant that where we can and I think here we can avoid it, by doing some simple traversal. I don't expect this code to get spread out over the entire code base. So handling type params here seems fine.